demo/Common/mimxrt1060/transport/using_mbedtls/using_mbedtls.c

```diff
@@ -801,39 +798,49 @@ int32_t TLS_FreeRTOS_send( NetworkContext_t * pNetworkContext,
                            size_t bytesToSend )
 {
     TlsTransportParams_t * pTlsTransportParams = NULL;
-    int32_t tlsStatus = 0;
+    size_t xWritten = 0;
+    int32_t lResult = 0;
 
     configASSERT( ( pNetworkContext != NULL ) && ( pNetworkContext->pParams != NULL ) );
 
     pTlsTransportParams = pNetworkContext->pParams;
-    tlsStatus = ( int32_t ) mbedtls_ssl_write( &( pTlsTransportParams->sslContext.context ),
-                                               pBuffer,
-                                               bytesToSend );
 
-    if( ( tlsStatus == MBEDTLS_ERR_SSL_TIMEOUT ) ||
-        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ||
-        ( tlsStatus == MBEDTLS_ERR_SSL_WANT_WRITE ) )
+    while( xWritten < bytesToSend )
     {
-        LogDebug( ( "Failed to send data. However, send can be retried on this error. "
-                    "mbedTLSError= %s : %s.",
-                    mbedtlsHighLevelCodeOrDefault( tlsStatus ),
-                    mbedtlsLowLevelCodeOrDefault( tlsStatus ) ) );
-
-        /* Mark these set of errors as a timeout. The libraries may retry send
-         * on these errors. */
-        tlsStatus = 0;
+        lResult = mbedtls_ssl_write( &( pTlsTransportParams->sslContext.context ),
+                                     pBuffer + xWritten,
+                                     bytesToSend - xWritten );
+
+        if( lResult > 0 )
+        {
+            /* Sent data, so update the tally and keep looping. */
+            xWritten += ( size_t ) lResult;
+        }
+        else if( ( lResult == 0 ) || ( lResult == -pdFREERTOS_ERRNO_ENOSPC ) )
+        {
+            /* No data sent. The secure sockets
+            * API supports non-blocking send, so stop the loop but don't
+            * flag an error. */
+        	lResult = 0;
```

Comment:
1 - I'm concerned on this - we break out of loop but we return to caller that we've written the `xWritten` which may not be accurate.  Is the socket guaranteed to write *all* the data or can it chose to just write some?

2 - Another question is on buffer lifetime management for caller?  After we return from `TLS_FreeRTOS_send` will the caller assume it doesn't need to keep buffer around anymore?  Will a non-blocking I/O make a copy of that buffer so this is OK, or is there risk of freeing?

3 - Also the `-pdFREERTOS_ERRNO_ENOSPC` - putting a negative in front of an error code - isn't a convention I've ever seen.  Please add a comment about what's up with this.

4 - My ;tl-dr here maybe should be "Can we (or should we) flip on blocking I/O and what happens if we do other than potentially some perf hit?"